File: ios/sensors/AccStreamManager.swift
import Foundation
import PolarBleSdk
import RxSwift
import React

class AccStreamManager {

    private weak var bridge: PolarBridge?
    private var api: PolarBleApi?

    private var accDisposable: Disposable?
    private var isAccStreaming = false

    private var accBuffer: [[String: Any]] = []
    private let accBufferQueue = DispatchQueue(label: "com.polarbridge.accBufferQueue")
    private var accFlushTimer: Timer?

    private let SENSOR_BUFFER_MS: TimeInterval = 10_000
    private let disposeBag = DisposeBag()

    init(api: PolarBleApi?, bridge: PolarBridge) {
        self.api = api
        self.bridge = bridge
    }

    // MARK: - Public

    func fetchAccData(_ deviceId: String, bufferMs: NSNumber?) {
        NSLog("AccStreamManager: Fetch ACC Data called on: \(deviceId) bufferMs: \(String(describing: bufferMs))")

        let resolvedBufferMs: TimeInterval
        if let number = bufferMs, !(number is NSNull) {
            resolvedBufferMs = number.doubleValue >= 0 ? number.doubleValue : SENSOR_BUFFER_MS
        } else {
            resolvedBufferMs = SENSOR_BUFFER_MS
        }

        guard let api = api else {
            NSLog("AccStreamManager: Polar API not initialized")
            return
        }

        if isAccStreaming {
            disposeAccStream()
            stopAccFlushTimer()
            flushAccBuffer()
            isAccStreaming = false

            bridge?.sendEvent(
                withName: PolarEvent.PolarAccComplete.rawValue,
                body: ["message": "ACC Stream stopped"]
            )
            return
        }

        isAccStreaming = true
        startAccFlushTimer(bufferMs: resolvedBufferMs)

        accDisposable = SensorSettings.requestStreamSettings(
                api: api,
                identifier: deviceId,
                feature: .acc
            )
            .flatMap { settings in
                api.startAccStreaming(deviceId, settings: settings).asObservable()
            }
            .observe(on: MainScheduler.instance)
            .subscribe(
                onNext: { [weak self] accData in
                    guard let self = self else { return }

                    self.accBufferQueue.async {
                        for sample in accData {
                            var event: [String: Any] = [:]
                            event["accX"] = sample.x
                            event["accY"] = sample.y
                            event["accZ"] = sample.z
                            event["accTimestamp"] = Double(sample.timeStamp)

                            self.accBuffer.append(event)
                        }
                    }
                },
                onError: { [weak self] error in
                    guard let self = self else { return }

                    self.stopAccFlushTimer()
                    self.flushAccBuffer()
                    self.isAccStreaming = false

                    self.bridge?.sendEvent(
                        withName: PolarEvent.PolarAccError.rawValue,
                        body: ["error": error.localizedDescription]
                    )

                    self.accDisposable = nil
                },
                onCompleted: { [weak self] in
                    guard let self = self else { return }

                    self.stopAccFlushTimer()
                    self.flushAccBuffer()
                    self.isAccStreaming = false

                    self.bridge?.sendEvent(
                        withName: PolarEvent.PolarAccComplete.rawValue,
                        body: ["message": "ACC stream complete"]
                    )

                    self.accDisposable = nil
                }
            )

        accDisposable?.disposed(by: disposeBag)
    }

    func disposeAccStream() {
        isAccStreaming = false
        accDisposable?.dispose()
    }

    // MARK: - Buffer

    private func startAccFlushTimer(bufferMs: TimeInterval) {
        stopAccFlushTimer()

        let intervalSeconds = bufferMs / 1000.0

        DispatchQueue.main.async {
            self.accFlushTimer = Timer(
                timeInterval: intervalSeconds,
                repeats: true
            ) { [weak self] _ in
                self?.flushAccBuffer()
            }

            RunLoop.main.add(self.accFlushTimer!, forMode: .common)
        }
    }

    private func stopAccFlushTimer() {
        DispatchQueue.main.async {
            self.accFlushTimer?.invalidate()
            self.accFlushTimer = nil
        }
    }

    private func flushAccBuffer() {
        accBufferQueue.async {
            guard !self.accBuffer.isEmpty else { return }

            let events = self.accBuffer
            self.accBuffer.removeAll()

            DispatchQueue.main.async {
                for event in events {
                    self.bridge?.sendEvent(
                        withName: PolarEvent.PolarAccData.rawValue,
                        body: event
                    )
                }
            }
        }
    }
}

File: android/src/main/java/com/polarbridge/sensors/AccStreamManager.kt
package com.polarbridge.sensors

import android.util.Log
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import com.polar.sdk.api.PolarBleApi
import com.polar.sdk.api.errors.PolarInvalidArgument
import com.polar.sdk.api.model.PolarSensorSetting
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import java.util.concurrent.TimeUnit

class AccStreamManager(
  private val api: PolarBleApi,
  private val sendEvent: (String, WritableMap) -> Unit,
  private val TAG: String = "AccStreamManager"
) {
  private var accDisposable: Disposable? = null

  fun fetchAccData(deviceId: String, sensorSettings: SensorSettings, ms: Double?, sensorBufferMs: Long) {
    Log.d(TAG, "Fetch Accelerometer Data called on: $deviceId ")
    val bufferMs = ms?.toLong()?.takeIf { it >= 0 } ?: sensorBufferMs
    val isDisposed = accDisposable?.isDisposed ?: true
    try{
      if (isDisposed) {
        accDisposable = sensorSettings.requestStreamSettings(deviceId, PolarBleApi.PolarDeviceDataType.ACC)
          .flatMap { settings: PolarSensorSetting ->
            api.startAccStreaming(deviceId, settings)
          }
          .flatMapIterable { it.samples }
          .buffer(bufferMs, TimeUnit.MILLISECONDS)
          .observeOn(AndroidSchedulers.mainThread())
          .subscribe(
            { samples ->
              if (samples.isEmpty()) return@subscribe
              Log.d(TAG, "Flushing ACC buffer (${samples.size} samples)")

              for (data in samples) {
                // Log.d(TAG, "ACC    x: ${data.x} y: ${data.y} z: ${data.z} timeStamp: ${data.timeStamp}")

                val event: WritableMap = Arguments.createMap()
                event.putInt("accX", data.x)
                event.putInt("accY", data.y)
                event.putInt("accZ", data.z)
                // Long not supported, use double as workaround
                // See: https://github.com/facebook/react-native/issues/9685
                event.putDouble("accTimestamp", data.timeStamp.toDouble())

                sendEvent("PolarAccData", event)
              }
            },
            { error: Throwable ->
              Log.e(TAG, "ACC stream failed. Reason $error")

              val errorEvent = Arguments.createMap()
              errorEvent.putString("error", error.toString())
              sendEvent("PolarAccError", errorEvent)
            },
            {
              Log.d(TAG, "ACC stream complete")

              val completeEvent = Arguments.createMap()
              completeEvent.putString("message", "ACC stream complete")
              sendEvent("PolarAccComplete", completeEvent)
            }
          )
      } else {
        // NOTE dispose will stop streaming if it is "running"
        disposeAccStream()
        Log.d(TAG, "ACC stream stopped")
      }
    } catch(polarInvalidArgument: PolarInvalidArgument){
      Log.e(TAG, "Failed to fetch ACC Data. Reason $polarInvalidArgument ")
    }
  }

  fun disposeAccStream(){
    try {
      accDisposable?.dispose()
    } catch (e: Exception) {
      Log.w(TAG, "Failed to dispose ACC stream cleanly. Bluetooth service likely died.", e)
    } finally {
      accDisposable = null
    }
  }
}
